### ISSUE TITLE
TypeScript Type Definitions: last mile?

### ISSUE THREAD

**Issue by altano** [mopidy/mopidy.js#27]
I was working on a project that relied on Mopidy.js so I started writing a TypeScript definition file for the API. It is mostly done but I have some more urgent matters come up and am not going to be able to finish the definitions and submit them to DefinitelyTyped.

If anyone wants to pick up where I left off, I think there's <10% left to do, and the hard parts like the `on`/`off` event subscription API are already done. Just look for "@TODO" comments in the code. Feel free to finish it and submit to DefinitelyTyped with your name instead of mine:

https://files.terriblefish.com/mopidyjs/index.d.ts

**Comment by jodal** [comment]
Cool!

I'm interested in finishing these and shipping them with Mopidy.js. I'm using TypeScript at work these days, so I should be able to figure out the remainder :-)

**Comment by altano** [comment]
Excellent @jodal! 

I’ll think you’ll want to ship them to DefinitelyTyped as there’s no point in maintaining the types here, unless the Mopidy.js team wants to own and maintain it themselves. I’ll of course defer to them.

Let me know if you have any questions about the types or need any help.

**Comment by altano** [comment]
Oh sorry you ARE the Mopidy team, hah! Well then please do as you please and just let me know if anything in the types thus far makes you scratch your head.

**Comment by whomwah** [comment]
I've needed this for a personal project that relies on `mopidy.js` too, and have tweaked/added the bits I think are missing. I'm still quite new to definition files so someone with more experience may need to tweak it further, but it's working OK for my use case.

https://gist.github.com/whomwah/61214417c98026232f19742e0ca666ec

**Comment by jodal** [comment]
I've combined the efforts by @altano and @whomwah here in #27 and @cristianpb in #33 into PR #36, which is now merged. It would be great if you all (@altano, @whomwah, @cristianpb, @LorenzKahl) could give the current defintions a spin and point out any issues before this is released to npm.

**Comment by LorenzKahl** [comment]
Hi @jodal, that's great news. Thank you very much for the quick reaction and the effort. I will tryout the type definitions the next few days and report back if I find something off or missing. 
Cheers, Lorenz

**Comment by cristianpb** [comment]
Thanks @jodal 
I just tested the type definitions locally and they are fine for me.

**Comment by LorenzKahl** [comment]
I'm still at the very start of my project as well as dipping my toes into Mopidy. So my usage of the API and thus the types is to be considered high level (toggling playback state / next track / previous track / clearing the playlist / storing the playhead on track change if not played until the end). 

So far everything worked like a charm for me. Having the types ready and using VS Code I found the API to be pretty much self explanatory (couple of months ago, when I first approached Mopidy.js I wouldn't say the same). So, this really helps a lot. 

Thank you very much for the inclusion!!!

PS: The only thing that got me tripping was, when installing via GitHub (`npm i mopidy/mopidy.js#master --save`) the types file (`node_modules/mopidy/src/mopidy.d.ts`) was not included - so I had to create/copy/paste it myself.

**Comment by altano** [comment]
Thanks for doing this, especially given that you hesitated to provide the type definitions in the first place. For what it's worth, the developer ergonomics of having these type definitions, ESPECIALLY because of the dynamic API, are super super worth it.

Anyway, I took a quick peak by manually copying your .d.ts file to `node_modules/mopidy/src/mopidy.d.ts`, just as @LorenzKahl did. Here's what I observed:

1. The generic "event" listener is gone. This is valid with mopidy but won't type check anymore:
```
mopidy.on("event", handleGenericEvent);
```
I was handling this in my code with this line, which I guess you could add to your `StrictEvents`, alongside `state: (args?: unknown) => void;`:
```
    /**
     * The events from Mopidy are also emitted under the aggregate event named
     * event.
     */
    event: (args?: unknown) => void;
```
This type isn't super important as the only useful way of using this kind of event listener is for generic console logging or debugging, since you're circumventing all the type safety. But it should be possible since it is in the Mopidy API.
2. I had to change from importing the `Models` type from the module to `models`. I assume that was an intentional fix and your version is correct.

Those were the only type errors I had to address after using your version. I did _not_ test the inverse: making sure that any type errors I might introduce into my code correctly trigger type errors with your version.

**Comment by whomwah** [comment]
Thanks for this @jodal ! I'd been using the definitions on a personal project https://github.com/kyan/jukebox-js (specifically `backend/src`) which was the motivation to finish of @altano's initial work. This was the first big definitions file I've looked so bow to other peoples experience here.

I'll move my manually created definitions file into the `node_modules/mopidy/src` later and check everything behaves OK.

**Comment by whomwah** [comment]
Removing my local `mopidy.d.ts` and downloading the version in `master` to `node_modules/mopidy/src/mopidy.d.ts` appears to work just fine for me too 👍

**Comment by jodal** [comment]
I did a few additional changes, including adding @altano's "event" type, and published v1.2.0 and v1.2.1 to npm. The latter version should include the TypeScript definitions.
